### PR TITLE
Feature/ch128/visualise variable state

### PR DESCRIPTION
**Pull request by claralebbos**
### Description

Added visualisation of variable state and code scrolling

Fixes # (issue) 

### Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Make sure variables are updated at the right time in current tests

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 1]
```diff
@@ -25,10 +25,12 @@ class VirtualMachine(
     private val displayLine: MutableList<Int> = mutableListOf()
     private val displayCode: MutableList<String> = mutableListOf()
     private val acceptableNonStatements = setOf("}", "{", "")
-    private val ALLOCATED_STACKS = 1000
+    private val ALLOCATED_STACKS = Runtime.getRuntime().freeMemory()/1000000
+    private val MAX_DISPLAYED_VARIABLES = 4
 
     init {
         fileLines.indices.forEach {
+            println(ALLOCATED_STACKS)
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:33]: Is this meant to be here? This is not relevant to this PR

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 2]
```diff
@@ -25,10 +25,12 @@ class VirtualMachine(
     private val displayLine: MutableList<Int> = mutableListOf()
     private val displayCode: MutableList<String> = mutableListOf()
     private val acceptableNonStatements = setOf("}", "{", "")
-    private val ALLOCATED_STACKS = 1000
+    private val ALLOCATED_STACKS = Runtime.getRuntime().freeMemory()/1000000
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:28]: Should be in another PR not relevant

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 3]
```diff
@@ -107,9 +152,12 @@ class VirtualMachine(
             linearRepresentation.add(Sleep(length))
         }
 
+        private fun moveToLine(line: Int = pc) {
+            linearRepresentation.add(MoveToLine(displayLine[line - 1], pointerVariable, codeBlockVariable, codeTextVariable))
+        }
         private fun moveToLine(line: Int = pc, updatePc: Boolean = false) {
             if (updatePc) pc = line
-            linearRepresentation.add(MoveToLine(displayLine[line - 1], pointerVariable, codeBlockVariable))
+            linearRepresentation.add(MoveToLine(displayLine[line - 1], pointerVariable, codeBlockVariable, codeTextVariable))
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:160]: Isn't this duplicated?

File: src/test/kotlin/com/manimdsl/ASTExecutorTests.kt [hunk 4]
```diff
@@ -10,33 +9,51 @@ import org.junit.jupiter.api.Test
...
+//            MoveToLine(lineNumber = 4, pointerName = "pointer", codeBlockName = "code_block", codeTextVariable = "code_text"),
+//            UpdateVariableState(variables= listOf("\"ans = 9.0\""), ident="variable_block", textColor=null),
+//            UpdateVariableState(variables= listOf("\"ans = 9.0\""), ident="variable_block", textColor=null),
+//            Sleep(length=1.0)
+//        )
+//        val actual = VirtualMachine(abstractSyntaxTree, symbolTable, lineNodeMap, program.split("\n"), Stylesheet(null, symbolTable)).runProgram().second
+//
+//        assertEquals(expected, actual)
+//
+//    }
```
Comment (arj119) [inline src/test/kotlin/com/manimdsl/ASTExecutorTests.kt:56]: Why is this commented out

**Comment by arj119** [comment]
Also fill out PR template and can you show an example animation of this working?

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 5]
```diff
@@ -42,10 +47,12 @@ class VirtualMachine(
     }
 
     fun runProgram(): Pair<ExitStatus, List<ManimInstr>> {
-        linearRepresentation.add(CodeBlock(displayCode, codeBlockVariable, codeTextVariable, pointerVariable))
+        linearRepresentation.add(PartitionBlock("1/3", "2/3"))
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:50]: Will this mess with the way the algorithm allocates space for data structures on the RHS? I thought the ratio was 5/14 : 9/14

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 6]
```diff
@@ -91,34 +140,38 @@ class VirtualMachine(
...
-                    stepInto = true
-                    EmptyValue
-                }
-                is StopStepIntoNode -> {
-                    stepInto = false
-                    EmptyValue
-                }
-                else -> EmptyValue
+        private fun getVariableState(): List<String>  {
+            return (0 until displayedDataMap.size).map { wrapString("${displayedDataMap[it]!!.first} = ${displayedDataMap[it]!!.second}") }
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:150]: `displayedDataMap.toSortedMap().values.map { wrapString("\"${it.first} = ${it.second}\"") }` might be a bit cleaner if you need the output in numerical order? Correct me if I'm missing something though

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt (hunk 5 above)
Comment (adityagoel4512) [reply src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:50]: I'm not 100% sure but it seems to look fine for programs that have been tried. The partition block makes frames for the different subsections so if those are being used hopefully it works fine

Comment (mayanksharma3) [reply src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:50]: Yeah I think because it's 1/3 the code block will actually be tiny bit smaller than it can be, but at least it's smaller over larger

**Review by rbanerjee20 — APPROVED** [review]
